src/v/cloud_storage/recovery_utils.cc

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/recovery_utils.h"
+
+#include "cloud_storage/logger.h"
+
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+namespace cloud_storage {
+
+ss::sstring
+generate_result_path(const storage::ntp_config& ntp_cfg, bool result) {
+    auto uuid = boost::uuids::random_generator()();
+    return fmt::format(
+      recovery_result_format,
+      recovery_result_prefix,
+      ntp_cfg.ntp().ns(),
+      ntp_cfg.ntp().tp.topic(),
+      ntp_cfg.ntp().tp.partition,
+      boost::uuids::to_string(uuid),
+      result);
+}
+
+ss::future<> place_download_result(
+  remote& remote,
+  cloud_storage_clients::bucket_name bucket,
+  const storage::ntp_config& ntp_cfg,
+  bool result_completed,
+  retry_chain_node& parent) {
+    retry_chain_node fib{&parent};
+    auto result_path = generate_result_path(ntp_cfg, result_completed);
+    auto result = co_await remote.upload_object(
+      bucket, cloud_storage_clients::object_key{result_path}, "", fib);
+    if (result != upload_result::success) {
+        vlog(
+          cst_log.error,
+          "failed to upload result to {}, error: {}",
+          result_path,
+          result);
+    }
+    co_return;
+}
+
+ss::future<std::vector<recovery_result>> gather_recovery_results(
+  remote& remote,
+  cloud_storage_clients::bucket_name bucket,
+  retry_chain_node& parent) {
+    retry_chain_node fib{&parent};
+    std::vector<recovery_result> results{};
+    auto result = co_await remote.list_objects(
+      bucket, fib, cloud_storage_clients::object_key{recovery_result_prefix});
+    if (result.has_error()) {
+        vlog(
+          cst_log.error, "failed to list recovery results: {}", result.error());
+        co_return results;
+    }
+
+    std::regex result_expr{fmt::format(
+      "{}/(.*?)/(.*?)/(\\d+)_(.*?).(true|false)", recovery_result_prefix)};
+
+    results.reserve(result.value().size());
+    std::for_each(
+      std::make_move_iterator(result.value().begin()),
+      std::make_move_iterator(result.value().end()),
```

Comment:
It doesn't make much sense to use move iterator here. The value is used to make a local copy of the key. And the copy of the key is passed to `std::regex_match` which accepts it by reference. It is possible to avoid copying by using `std::move` on line 77, but you still can't move `ss::sstring` on line 76.  

If you want to avoid making copies you can use different overload of the `std::regex_match` which accepts string by two iterators. It will be something like `if (std::regex_match(item.key.begin(), item.key.end(), matches, result_expr))` on line 77. And you don't need to use `std::for_each` and `std::make_move_iterator` in this case. Simple for loop will do.

Reply:
I had tried earlier to use the std::regex_match overload with iterator, also tried to pass it a string_view so the const char* overload would be used, for some reason it always fails to compile unless it is given a string, eg with the iterators:

```
/development/redpanda/src/v/cloud_storage/recovery_utils.cc:85:13: error: no matching function for call to 'regex_match'
        if (std::regex_match(
            ^~~~~~~~~~~~~~~~
/development/redpanda/vbuild/llvm/install/bin/../include/c++/v1/regex:6214:1: note: candidate template ignored: deduced conflicting types for parameter '_BidirectionalIterator' ('const char *' vs. 'std::__wrap_iter<const char *>')
regex_match(_BidirectionalIterator __first, _BidirectionalIterator __last,
^
/development/redpanda/vbuild/llvm/install/bin/../include/c++/v1/regex:6235:1: note: candidate template ignored: could not match 'basic_regex' against 'match_results'
regex_match(_BidirectionalIterator __first, _BidirectionalIterator __last,
^
/development/redpanda/vbuild/llvm/install/bin/../include/c++/v1/regex:6246:1: note: candidate template ignored: could not match 'match_results<const _CharT *, _Allocator>' against 'const char *'
regex_match(const _CharT* __str, match_results<const _CharT*, _Allocator>& __m,
^
/development/redpanda/vbuild/llvm/install/bin/../include/c++/v1/regex:6256:1: note: candidate template ignored: could not match 'basic_string<_CharT, _ST, _SA>' against 'const char *'
regex_match(const basic_string<_CharT, _ST, _SA>& __s,
^
/development/redpanda/vbuild/llvm/install/bin/../include/c++/v1/regex:6268:1: note: candidate template ignored: could not match 'basic_string<_CharT, _ST, _SA>' against 'const char *'
regex_match(const basic_string<_CharT, _ST, _SA>&& __s,
^
/development/redpanda/vbuild/llvm/install/bin/../include/c++/v1/regex:6277:1: note: candidate function template not viable: requires at most 3 arguments, but 4 were provided
regex_match(const _CharT* __str, const basic_regex<_CharT, _Traits>& __e,
^
/development/redpanda/vbuild/llvm/install/bin/../include/c++/v1/regex:6286:1: note: candidate function template not viable: requires at most 3 arguments, but 4 were provided
regex_match(const basic_string<_CharT, _ST, _SA>& __s,
^
```

The call only seems to compile with a `std::string` as the first parameter. I did change the code to a simple loop over const references to the list result, there is one copy of the `item.key` but I couldn't find a way around it.

Reply:
fixed